Source/Dafny/Resolver.cs

```diff
@@ -12510,12 +12545,13 @@ private class RBranchExpr : RBranch {
         Console.WriteLine("DEBUG: CompileNestedMatchExpr for match at line {0}", e.tok.line);
       }
 
-      MatchTempInfo mti = new MatchTempInfo(e.tok, e.Cases.Count, opts.codeContext, DafnyOptions.O.MatchCompilerDebug);
+      var cases = e.Cases.SelectMany(FlattenNestedMatchCaseExpr).ToList();
```

Comment:
What is the right time for flattening the disjunctive patterns? Can't that be done before resolution as a separate rewriter?

Reply:
Possibly, but I'm not sure it's what I want: I wrote it this way to give other code access to the disjunctive patterns.  I think of this flattening as an implementation detail of the pattern compiler.

Reply:
> I wrote it this way to give other code access to the disjunctive patterns.

Didn't we want to use simple lowering transformations when possible, to make each individual transformation simpler? I think it would be a big benefit if we can avoid adding code to `Resolver.cs`.

Also, in what scenario do you envision someone needs access to disjunctive patterns? And if they need it, why couldn't they sit before the lowering transformation?

Reply:
> Didn't we want to use many small lowering transformations when possible, to make each individual transformation simpler?

Indeed, and we have that here, right?  There's just a handful of calls in the resolver to new code that pre-processes the pattern.

> I think it would be a big benefit if we can avoid adding code to Resolver.cs.

I can move it to a new file, but all of the other pattern matching compilation code is there, so… ?

Originally, the pattern matching compiler was implemented all in the parser.  It was moved to the resolver, and in fact it should really be moved to the compiler.  In fact, in the long run, disjunctive patterns should be supported at any level of nesting, so the nice separation we have right now won't survive.  But for now it's good to have the separated implementation.

Reply:
> Indeed, and we have that here, right? There's just a handful of calls in the resolver to new code that pre-processes the pattern.
> I can move it to a new file, but all of the other pattern matching compilation code is there, so… ?

I meant moving the code to a separate file and not calling it from Resolver.cs, except for possibly adding a `DisjunctCaseRewriter` to the `rewriters` list. Would that be possible?

> In fact, in the long run, disjunctive patterns should be supported at any level of nesting, so the nice separation we have right now won't survive. But for now it's good to have the separated implementation.

I don't know the pattern matching transformations well enough to comment.

Reply:
> I meant moving the code to a separate file and not calling it from Resolver.cs, except for possibly adding a DisjunctCaseRewriter to the rewriters list. Would that be possible?

Possibly?  I *think* it might work, because this pass doesn't use type information.  The problem is that pattern matching compilation happens in the middle of resolution, and it uses some type info.

Also, do we have the architecture in place for writing such a pass?  I need something that traverses all statements and expressions in a program.  Do we have a visitor that does that?

Reply:
Editing to add this following our conversation: it would be nice to add such a visitor, but one issue with rewriting this systematically would be that clients would lose access to the unresolved AST containing the full match structure.